spec-smt/ssf.smt

```diff
@@ -0,0 +1,327 @@
+            (< (checkpoint_slot (source vote1)) (checkpoint_slot (source vote2)))
+            (and
+                ; vote1.source.chkp_slot = vote2.source.chkp_slot
+                (= (checkpoint_slot (source vote1)) (checkpoint_slot (source vote2)))
+                ; vote1.source.block.slot < vote2.source.block.slot
+                (< (slot (checkpoint_block (source vote1))) (slot (checkpoint_block (source vote2))))
+            )
+        )
+        ; vote2.target.chkp_slot < vote1.target.chk_slot
+        (< (checkpoint_slot (target vote2)) (checkpoint_slot (target vote1)))
+    )
+)
+
+; slashable nodes
+(declare-const slashable_nodes (Set Node))
+(assert (= slashable_nodes (set.comprehension ((node Node))
+        (exists ((vote1 Vote) (vote2 Vote))
+            (and
+                (= (sender vote1) node)
+                (= (sender vote2) node)
+                (set.member vote1 votes)
+                (set.member vote2 votes)
+                (or
+                    (are_equivocating_votes vote1 vote2)
+                    (does_first_vote_surround_second_vote vote1 vote2)
+                )
+            )
+        )
+        node
+    )
+))
+
+; ============================
+; Examples
+; ============================
+
+; uncomment one of the following queries to search for a model
+
+; ; find conflicting blocks
+; (assert (not (= conflicting_blocks (as set.empty (Relation Hash Hash)))))
+; (check-sat)
+; (get-model)
+
+; ; find a justified checkpoint (besides genesis)
+; (assert (not (= justified_checkpoints (set.singleton genesis_checkpoint))))
+; (check-sat)
+; (get-model)
+
+; ; find a finalized checkpoint (besides genesis)
+; (assert (not (= finalized_checkpoints (set.singleton genesis_checkpoint))))
+; (check-sat)
+; (get-model)
+
+; ; find two conflicting finalized blocks
+; (assert (exists ((block1 Hash) (block2 Hash))
+;     (and
+;         (set.member (tuple block1 block2) conflicting_blocks)
+;         (set.member block1 finalized_blocks)
+;         (set.member block2 finalized_blocks)
+;     )
+; ))
+; (check-sat)
+; (get-model)
+
+; ; find a slashable node
+; (assert (not (= slashable_nodes (as set.empty (Set Node)))))
+; (check-sat)
+; (get-model)
+
+; ============================
+; Accountable Safety
+; ============================
+
+; We want to check
+;     \E b1, b2 : b1 and b2 are conflicting and finalized  => \E node : node is slashable
+;
+; This formula is valid iff its negation
+;     \E b1, b2 : b1 and b2 are conflicting and finalized  /\  ~ (\E node : node is not slashable)
+; is unsatisfiable.
+(assert
+    (and
+        (exists ((block1 Hash) (block2 Hash))
+            (and
+                (set.member (tuple block1 block2) conflicting_blocks)
+                (set.member block1 finalized_blocks)
+                (set.member block2 finalized_blocks)
+            )
+        )
+        (not (exists ((node Node)) (set.member node slashable_nodes)))
```

Comment:
Would it be cheaper to write this quantifier over finalized checkpoints (and defining `conflicting_checkpoints` analogously) and map the blocks out? You could avoid axiomatizing `finalized_blocks`.

Reply:
True, I tried and it's slower for the two smaller benchmark parameters though.
So let's keep it as is.